File: c++/14621.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

char univ[1001];
vector<pair<int, pair<int, int>>> routes;
int parents[1001];

int getParent(int x) {
  if (x == parents[x])
    return x;
  else
    return parents[x] = getParent(parents[x]);
}

void unionParent(int a, int b) {
  a = getParent(a);
  b = getParent(b);

  if (a > b)
    parents[a] = b;
  else if (a < b)
    parents[b] = a;
}

bool sameParent(int a, int b) {
  a = getParent(a);
  b = getParent(b);

  return a == b;
}

int main() {
  freopen("example.txt", "r", stdin);

  int N, M;
  cin >> N >> M;

  for (int i = 1; i <= N; i++) {
    cin >> univ[i];
    parents[i] = i;
  }

  for (int i = 0; i < M; i++) {
    int u, v, d;
    cin >> u >> v >> d;

    routes.push_back({d, {u, v}});
  }

  sort(routes.begin(), routes.end());

  int cnt = 0;
  int ans = 0;

  for (int i = 0; i < M; i++) {
    int weight = routes[i].first;
    int from = routes[i].second.first;
    int to = routes[i].second.second;

    if (!sameParent(from, to) && univ[from] != univ[to]) {
      unionParent(from, to);
      ans += weight;
      cnt++;
    }

    if (cnt == N - 1) break;
  }

  cout << ((cnt == N - 1) ? ans : -1);

  return 0;
}

File: js/14621.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

let idx = 0;
const getLine = () => inputs[idx++];

const [N, M] = getLine().split(' ').map(Number);

const univ = getLine().split(' ');
const routes = [];
const parents = Array(N + 1)
  .fill()
  .map((_, i) => i);

for (let i = 0; i < M; i++) {
  const route = getLine().split(' ').map(Number);

  routes.push(route);
}

routes.sort((a, b) => a[2] - b[2]);

let cnt = 0;
let ans = 0;

for (let i = 0; i < M; i++) {
  const [from, to, weight] = routes[i];

  if (!sameParent(from, to) && univ[from - 1] !== univ[to - 1]) {
    unionParent(from, to);
    ans += weight;
    cnt++;
  }

  if (cnt === N - 1) break;
}

console.log(cnt === N - 1 ? ans : -1);

function getParent(x) {
  if (x === parents[x]) return x;
  else return (parents[x] = getParent(parents[x]));
}

function unionParent(a, b) {
  a = getParent(a);
  b = getParent(b);

  if (a < b) parents[b] = a;
  else if (a > b) parents[a] = b;
}

function sameParent(a, b) {
  a = getParent(a);
  b = getParent(b);

  return a === b;
}
